### ISSUE TITLE
☂️ Move to on-demand Windows API projection

### ISSUE THREAD

**Issue by halildurmus** [halildurmus/win32#819]
## Background

`package:win32` currently ships a pre-generated subset of the Windows API, which covers many common use cases. However, this approach doesn't scale well — the Win32 API surface is enormous, and broadening coverage means unconditionally loading far more Dart code than most users will ever need, placing unnecessary strain on the Dart analyzer.

## Proposed Direction

Move to an **on-demand projection** model, similar to how `ffigen` works:

- `package:win32` is slimmed down to expose only a small core of helpers and runtime pieces.
- A new **generator package** (e.g., `package:win32gen`) accepts user input declaring which Win32 APIs they need, then emits only the corresponding Dart bindings into their project.
- The generator could also support customization around naming conventions, projection rules, and filtering.

## Benefits

- **Faster analyzer performance** — only the APIs a project actually uses are materialized as Dart code.
- **Smaller dependency footprint** — users aren't paying the cost for APIs they'll never call.
- **Improved discoverability** — generating bindings on demand makes the API selection explicit and documented in the project's config, making it clear to contributors which Win32 APIs a project depends on.
- **More flexibility** — projection rules and naming can be customized per-project.

---
<details>
<summary><b>Original issue description</b></summary>

~~Currently, the [JSON files](https://github.com/halildurmus/win32/tree/main/packages/generator/data) define which APIs are projected by the [generator](https://github.com/halildurmus/win32/tree/main/packages/generator).~~

~~In a future major release (not in `v6`, likely `v7`), I plan to expand the `generator` to project nearly the entire Win32 API, excluding a few outdated APIs. This will be made possible by leveraging Microsoft’s metadata, eliminating the need for the manual effort currently required to add new APIs -- a process that is both time-consuming and tedious.~~

~~The expanded scope will include:~~
~~- **Win32 callbacks**~~
~~- **Win32 constants**~~
~~- **Win32 enumerations**~~  
~~- **Win32 functions**~~
~~- **Win32 structures**~~
~~- **COM interfaces**~~

</details>

**Comment by Levi-Lesches** [comment]
I have a concrete use-case at https://github.com/YehudaKremer/msix/issues/292: [Determining if your app is running with package identity](https://learn.microsoft.com/en-us/windows/msix/detect-package-identity): 

```dart
#include <windows.h>
#include <appmodel.h>
#include <atlbase.h>
#include <VersionHelpers.h>

std::optional<bool> NativePlugin::checkIdentity() {
  if (!IsWindows8OrGreater()) return false;
  uint32_t length = 0;
  auto error = GetCurrentPackageFullName(&length, nullptr);
  if (error == APPMODEL_ERROR_NO_PACKAGE) {
    return false;
  } else if (error != ERROR_INSUFFICIENT_BUFFER) {
    return std::nullopt;
  }
  std::vector<wchar_t> fullName(length);
  error = GetCurrentPackageFullName(&length, fullName.data());
  if (error != ERROR_SUCCESS) return std::nullopt;
  return true;
}
```

Seems this package doesn't include `<appmodel.h>`, so I can't use `IsWindows8OrGreater()` or `GetCurrentPackageFullName()`.

**Comment by halildurmus** [comment]
Thanks for explaining your use case, @Levi-Lesches! I can add these functions easily.

**Comment by knopp** [comment]
That would be amazing. I'm currently in need for DirectComposition API and deciding whether I want to generate them or write a simply c wrapper for now.

One thing to consider - not sure what the policy is with umbrella import, but this is going to be *a lot* of Dart code and likely very heavy on analyzer if everything was unconditionally imported every time.

**Comment by halildurmus** [comment]
> One thing to consider - not sure what the policy is with umbrella import, but this is going to be _a lot_ of Dart code and likely very heavy on analyzer if everything was unconditionally imported every time.

I actually prototyped this and ended up at the same conclusion. The Win32 API surface is huge, and pulling everything in unconditionally would be rough on the analyzer and overkill for most users.

A better direction is probably on-demand projection, similar to how `ffigen` works. `package:win32` would expose a small core of helpers and runtime pieces, and a separate generator package would take user input (which APIs to generate) and emit only what they need. That generator could also support some customization around naming, projection rules, and so on.

**Comment by knopp** [comment]
I think that'd be great. Like for what I'm doing now, I need a small subset of API (Direct Composition + DXGI + Some bits of D3D11 and 12), would be nice to be able to specify which interfaces I need and the generator would figure out transitive dependencies and would make it happen.

**Comment by Levi-Lesches** [comment]
Wouldn't separate packages for different areas of the win32 API be simpler and still offer the same benefits?

**Comment by halildurmus** [comment]
> Wouldn't separate packages for different areas of the win32 API be simpler and still offer the same benefits?

I don't think so.

Win32 API surface is _huge_. Based on the latest [win32metadata](https://github.com/microsoft/win32metadata), we’re talking about:
 - Constants: **102,058**
 - Delegates: **4,030**
 - Enums: **9,893**
 - Functions: **20,357**
 - Structs: **17,894**
 - COM Interfaces: **16,011**

Even if we split APIs by top-level namespace (which would be ~16 packages), a few problems show up quickly:
- The namespaces aren’t evenly distributed, so some packages would be massive while others would be tiny.
- There are lots of shared types and cross-dependencies, which would inevitably introduce cyclic dependencies.
- Real-world usage would still require importing multiple packages, so even simple scenarios would put significant pressure on the analyzer.

**Comment by knopp** [comment]
Yeah, I'd definitely prefer generating the bindings on-demand, possibly with a rather strict whitelist (i.e. being able to disable certain COM members to prevent transitive dependencies).
